src/Microsoft.ML.Data/MLContext.cs

```diff
@@ -104,12 +108,25 @@ private void ProcessMessage(IMessageSource source, ChannelMessage message)
             log(this, new LoggingEventArgs(msg));
         }
 
-        bool IHostEnvironment.IsCancelled => _env.IsCancelled;
         string IExceptionContext.ContextDescription => _env.ContextDescription;
         TException IExceptionContext.Process<TException>(TException ex) => _env.Process(ex);
-        IHost IHostEnvironment.Register(string name, int? seed, bool? verbose) => _env.Register(name, seed, verbose);
+        IHost IHostEnvironment.Register(string name, int? seed, bool? verbose)
+        {
+            var host = _env.Register(name, seed, verbose);
+            _hosts.Add(host);
+            return host;
+        }
+
         IChannel IChannelProvider.Start(string name) => _env.Start(name);
         IPipe<TMessage> IChannelProvider.StartPipe<TMessage>(string name) => _env.StartPipe<TMessage>(name);
         IProgressChannel IProgressChannelProvider.StartProgressChannel(string name) => _env.StartProgressChannel(name);
+
+        [BestFriend]
+        internal void StopExecution()
+        {
+            foreach(var host in _hosts)
```

Comment:
Why is remembering the `_hosts` necessary?  Why can't we just call `_env.StopExecution()` here?

Reply:
StopExecution is implemented by HostBase not EnvironmentBase.

---
In reply to: [264861094](https://github.com/dotnet/machinelearning/pull/2797#discussion_r264861094) [](ancestors = 264861094)

Reply:
This doesn't make sense to me. The `HostEnvironmentBase` already remembers all of its `_children`. We should put the `StopExecution()` there, and MLContext doesn't need to do any of this. It can just call `_env.StopExecution()`.

---
In reply to: [264862395](https://github.com/dotnet/machinelearning/pull/2797#discussion_r264862395) [](ancestors = 264862395,264861094)